syntax = "proto3";

package protocol;

enum ProtocolVersionError {
  VERSION_OK = 0;
  CLIENT_TOO_NEW = 1;
  CLIENT_TOO_OLD = 2;
}

message ProtocolVersionRequest {
  uint32 protocol_version = 1;
}

message ProtocolVersionResponse {
  uint32 protocol_version = 1;
  ProtocolVersionError error = 2;
  string error_message = 3;
}

enum NumberSequenceError {
  SEQUENCE_OK = 0;
  INVALID_UPPER_BOUND = 1;
}

message NumberSequenceRequest {
  double upper_bound = 1;
  uint64 number_count = 2;
}

message NumberSequenceResponse {
  uint64 number_count = 1;
  double upper_bound = 2;
  uint64 sequence_index = 3;
  uint64 sequence_count = 4;
  uint64 sequence_number_count = 5;
  repeated double numbers = 6;
  uint64 checksum = 7;
  NumberSequenceError error = 8;
  string error_message = 9;
}

enum NumberSequenceAck {
  ACK_OK = 0;
  ACK_INVALID = 1;
}

message NumberSequenceAckRequest {
  uint64 sequence_index = 1;
  NumberSequenceAck ack = 2;
  uint64 checksum = 3;
}
